kernel/posix/mqueue.c

```diff
@@ -0,0 +1,532 @@
+	block = &mqd->block;
+	k_mem_pool_free(block);
+	k_sem_take(&mq_sem, K_FOREVER);
+	msg_queue->ref_count -= 1;
+	k_sem_give(&mq_sem);
+	return 0;
+}
+
+/**
+ * @brief Remove a message queue.
+ *
+ * See IEEE 1003.1
+ */
+int mq_unlink(const char *name)
+{
+	mqueue_object  *msg_queue;
+	mqueue_object  *next_mqueue, *prev_mqueue;
+	struct k_mem_block *block;
+
+	k_sem_take(&mq_sem, K_FOREVER);
+	msg_queue = find_in_list(name);
+
+	if (msg_queue != NULL && msg_queue->ref_count == 0) {
+		/* If he message queue is in the list, remove it. */
+		if (mq_list == msg_queue) {
+			mq_list = msg_queue->next;
+		} else {
+			prev_mqueue = msg_queue->prev;
+			next_mqueue = msg_queue->next;
+
+			if (prev_mqueue) {
+				prev_mqueue->next = next_mqueue;
+			}
+
+			if (next_mqueue) {
+				next_mqueue->prev = prev_mqueue;
+			}
+		}
+
+		k_msgq_purge(&msg_queue->queue);
+		msg_queue->next = msg_queue->prev = NULL;
+		k_sem_give(&mq_sem);
+
+		/* Free up name */
+		block = &msg_queue->block_buffer_name->block_name;
+		k_mem_pool_free(block);
+
+		/* Free up buffer track */
+		block = &msg_queue->block_buffer_name->block_buffer;
+		k_mem_pool_free(block);
+
+		/* Free up Queue object */
+		block = &msg_queue->block_mqueue;
+		k_mem_pool_free(block);
+
+		return 0;
+	}
+
+	k_sem_give(&mq_sem);
+	errno = EBADF;
+	return -1;
+}
+
+/**
+ * @brief Send a message to a message queue.
+ *
+ * See IEEE 1003.1
+ */
+int mq_send(mqd_t mqdes, const char *msg_ptr, size_t msg_len,
+	    unsigned int msg_prio)
+{
+	mqueue_desc *mqd = (mqueue_desc *)mqdes;
+	mqueue_object  *msg_queue;
+	s32_t  timeout = K_FOREVER, ret = -1;
+
+	if (mqd == NULL) {
+		errno = EBADF;
+		return ret;
+	}
+
+	msg_queue = mqd->mqueue;
+
+	if (mqd->flags & O_NONBLOCK) {
+		timeout = K_NO_WAIT;
+	}
+
+	if (k_msgq_put(&msg_queue->queue, (void *)msg_ptr, timeout) != 0) {
+		errno = EAGAIN;
+	} else {
+		ret = 0;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Receive a message from a message queue.
+ *
+ * See IEEE 1003.1
+ */
+int mq_receive(mqd_t mqdes, char *msg_ptr, size_t msg_len,
+		   unsigned int *msg_prio)
+{
+	mqueue_desc *mqd = (mqueue_desc *)mqdes;
+	mqueue_object  *msg_queue = mqd->mqueue;
+	s32_t  timeout = K_FOREVER, ret = -1;
+
+	if (mqd == NULL) {
+		errno = EBADF;
+		return ret;
+	}
+
+	/* Full message has not been received */
+	if (msg_len < msg_queue->queue.msg_size) {
+		errno = EAGAIN;
+		return ret;
+	}
+
+	if (mqd->flags & O_NONBLOCK) {
+		timeout = K_NO_WAIT;
+	}
+
+	if (k_msgq_get(&msg_queue->queue, (void *)msg_ptr, timeout) == 0) {
+		ret = msg_queue->queue.msg_size;
+	} else {
+		errno = EAGAIN;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Receive message from a message queue within abstime time.
+ *
+ * See IEEE 1003.1
+ */
+int mq_timedreceive(mqd_t mqdes, char *msg_ptr, size_t msg_len,
+			unsigned int *msg_prio, const struct timespec *abstime)
+{
+	mqueue_desc *mqd = (mqueue_desc *)mqdes;
+	mqueue_object  *msg_queue = mqd->mqueue;
+	s32_t  timeout = K_NO_WAIT, ret = -1;
+
```

Comment:
You're duplicating a ton of code here form mq_receive().  Please unify.  Code bytes aren't cheap.